networkx/algorithms/euler.py

```diff
@@ -147,3 +158,203 @@ def eulerian_circuit(G, source=None):
             arbitrary_edge = next(edges(current_vertex))
             vertex_stack.append(get_vertex(arbitrary_edge))
             g.remove_edge(*arbitrary_edge)
+
+
+def is_semieulerian(G):
+    """Returns ``True`` if and only if ``G`` has an Eulerian path.
+
+    An Eulerian path is a path that crosses every edge in G
+    exactly once.
+
+    Parameters
+    ----------
+    G: NetworkX Graph, DiGraph, MultiGraph or MultiDiGraph
+        A directed or undirected Graph or MultiGraph.
+
+    Returns
+    -------
+    True, False
+
+    See Also
+    --------
+    is_eulerian()
+
+    Examples
+    --------
+    >>> G = nx.DiGraph([(1,2),(2,3)])
+    >>> nx.is_semieulerian(G)
+    True
+    """
+    is_directed = G.is_directed()
+
+    # Verify that graph is connected, short circuit
+    if is_directed and not nx.is_weakly_connected(G):
+        return False
+
+    # is undirected
+    if not is_directed and not nx.is_connected(G):
+        return False
+
+    # Not all vertex have even degree, check if exactly two vertex
+    # have odd degrees.  If yes, then there is an Euler path. If not,
+    # raise an error (no Euler path can be found)
+
+    # if the odd condition is not meet, raise an error.
+    start = _find_path_start(G)
+    if not start:
+        return False
+
+    return True
+
+
+def _find_path_start(G):
+    """Returns a suitable starting vertex for Eulerian path
+
+    The function also verifies that the graph contains a path. If no
+    path exists, returns ``None''.
+
+    """
+    is_directed = G.is_directed()
+
+    # list to check the odd degree condition
+    check_odd = []
+
+    if is_directed:
+        degree = G.in_degree
+        out_degree = G.out_degree
+    else:
+        degree = G.degree
+
+    # Verify if an Euler path can be found. Complexity O(|V|)
+    for v in G:
+        deg = degree(v)
+        # directed case
+        if is_directed:
+            outdeg = out_degree(v)
+            if deg != outdeg:
+                # if we have more than 2 odd nodes no Euler path exists
+                if len(check_odd) > 2:
+                    return False
+                # is odd and we append it.
+                check_odd.append(v)
```

Comment:
Why are you maintaining a list of all odd degree nodes? It seems you only need the first one or two.